### PR TITLE
Deny create_translation_test if there exists failed test within last …

### PR DESCRIPTION
**Pull request by jennifer-lu**
…30 days

## Notion ticket link

[Deny create_translation_test if there exists failed test within last 30 days](https://www.notion.so/uwblueprintexecs/Deny-create_translation_test-if-there-exists-failed-test-within-last-30-days-87a44734a2ff45fbbb055ada4e6834ad)

## Implementation description

- Raise exceptions in `create_translation_test` service if the user has test in the language that:
    - is not published,
    - or has current time - translator_last_activity < 30 days <=> current time - 30 days < translator_last_activity

## Steps to test

1. Setup Altair as an admin
2. View `story_translations` table using mysql
3. Create a story translation test using `createStoryTranslationTest` mutation in Altair:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
4. The story translation test should appear in the `story_translations` table
5. Create a story translation test with the same user and language (level can be different) using `createStoryTranslationTest` mutation:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
6. Should get "User has an ongoing story translation test for language {language}." error message, and the story translation test should NOT appear in the `story_translations` table
7. Give a failing test grade using `finishGradingStoryTranslation` mutation:
```
mutation {
  finishGradingStoryTranslation(storyTranslationTestId: 14, testFeedback: "blah", testResult: "{}"){
    ok
    storyTranslation{
      id
      stage
      
    }
  }
}
```
8. Create a story translation test with the same user and language using `createStoryTranslationTest` mutation:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
9. Should get "User has failed a test within the last 30 days." error message, and the story translation test should NOT appear in the `story_translations` table
10. Set the created translation test reviewer_last_activity to more than 30 days ago using mysql `UPDATE story_translations SET reviewer_last_activity = '2021-11-01 00:00:00' WHERE id = 14;`
11. Create a story translation test with the same user and language using `createStoryTranslationTest` mutation:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
12. The story translation test should appear in the `story_translations` table
13. Create a story translation test with the same user and language using `createStoryTranslationTest` mutation:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
14.  Should get "User has an ongoing story translation test for language {language}." error message, and the story translation test should NOT appear in the `story_translations` table
15. Give a passing test grade using `finishGradingStoryTranslation` mutation:
```
mutation {
  finishGradingStoryTranslation(storyTranslationTestId: 15, testFeedback: "good", testResult: "{\"translate\": \"FRENCH\", \"review\": \"FRENCH\", \"asdfasf\": \"ASDFASDF\"}"){
    ok
    storyTranslation{
      id
      stage
      
    }
  }
}
```
16. Create a story translation test with the same user and language using `createStoryTranslationTest` mutation:
```
mutation { createStoryTranslationTest(userId: 1, level: 2, language: "GERMAN") {
      story {
        id
      }
    }
  }
```
17. The story translation test should appear in the `story_translations` table

## What should reviewers focus on?

- Does testing work
- Are exceptions correct
    - ~~To check if a test is ongoing, we look at `story_translation_test.stage != PUBLISHED` (should we instead look at something like `test_result == NULL`?)~~
    - ~~To check if there exists failed test within last 30 days, we look at current time - 30 days < translator_last_activity (is translator_last_activity the correct timestamp? should we also verify that `test_result == failed`?)~~

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters: `docker exec -it planet-read_py-backend_1 /bin/bash -c "black . && isort --profile black ."` and I have generated new migrations: `flask db migrate -m "<your message>"`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/python/app/services/implementations/story_service.py
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:134]: nit: `User has an ongoing story translation test for language {language}.`

**Review by gaoxk — CHANGES_REQUESTED** [review]
behaved as expected while testing! However, two thoughts. First:
> should we also verify that test_result == failed

yes, we should make sure the test is failed! Shoot, I didn't this while discussing the `FinishGradingStoryTranslation` mutation. Can you make some changes to the `finish_grading_story_translation` service function to handle this?
imo for `finish_grading_story_translation`,  if `test_result` lacks the translate key, then the test should be treated as a fail and `test_result` should be saved as `{}` in the database. ie at [line 687 of story_service, ](https://github.com/uwblueprint/planet-read/pull/199/files#diff-657b40175f47813e4a497d750fe834d3445eb33e0d27795f83bb1616345bf6daR687) consider the test failed rather than throwing an exception. 

Then, for your pr, the error `"User has failed a test within the last 30 days."` should throw after a recent test has been failed. [Here's ](http://localhost:5000/graphql?variables=null&operationName=undefined&query=mutation%20%7B%0A%20%20finishGradingStoryTranslation(storyTranslationTestId%3A%2014%2C%20testFeedback%3A%20%22blah%22%2C%20testResult%3A%20%22%7B%7D%22)%7B%0A%20%20%20%20ok%0A%20%20%20%20storyTranslation%7B%0A%20%20%20%20%20%20id%0A%20%20%20%20%20%20stage%0A%20%20%20%20%20%20%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A)an example failing grade mutation call, though note you'll have to test it in altair with an admin token.

Second thought: now that your `_update_story_translation_last_activity` ticket is done, I think it makes sense to modify `finish_grading_story_translation` to update the `reviewer_last_activity` column (since the admin is the reviewer). Then, rather than checking against 30 days since `translator_last_activity`, we should check 30 days since  `reviewer_last_activity`. (ie change lines 137-138 in story_service on this pr). 

I'm fine with either completing these in this pr or in a separate pr, though we'd need to merge the separate prs before merging this one. I'll leave it up to you to decide! LMK what you think and if you have any questions

File: backend/python/app/services/implementations/story_service.py
Comment (jennifer-lu) [inline backend/python/app/services/implementations/story_service.py:754]: Moved this to here, because it follows from `"translate" in test_result`.

**Review by gaoxk — APPROVED** [review]
nice!!! 🚀

**Pull request merged**
